archunit/src/main/java/com/tngtech/archunit/library/modules/ArchModules.java

```diff
@@ -0,0 +1,556 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.tngtech.archunit.library.modules;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.InvocationTargetException;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import com.google.common.base.Joiner;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.SetMultimap;
+import com.google.common.collect.Sets;
+import com.tngtech.archunit.PublicAPI;
+import com.tngtech.archunit.base.ForwardingCollection;
+import com.tngtech.archunit.core.domain.JavaClass;
+import com.tngtech.archunit.core.domain.JavaClasses;
+import com.tngtech.archunit.core.domain.PackageMatcher;
+import com.tngtech.archunit.library.dependencies.Slices;
+import com.tngtech.archunit.library.modules.ArchModule.Identifier;
+import com.tngtech.archunit.library.modules.ArchModules.Creator.WithGenericDescriptor;
+
+import static com.google.common.base.Functions.toStringFunction;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.collect.Multimaps.asMap;
+import static com.tngtech.archunit.PublicAPI.Usage.ACCESS;
+import static com.tngtech.archunit.PublicAPI.Usage.INHERITANCE;
+import static com.tngtech.archunit.core.domain.PackageMatcher.TO_GROUPS;
+import static com.tngtech.archunit.library.modules.ArchModule.Identifier.ignore;
+import static java.util.function.Function.identity;
+import static java.util.stream.Collectors.toMap;
+
+/**
+ * A collection of {@link ArchModule "architectural modules"}. This class provides a convenient API to partition the {@link JavaClass classes}
+ * of a code base into (cohesive) modules and assert properties of these modules. E.g. their dependencies to each other or dependencies
+ * not contained in any of the constructed modules.<br>
+ * This class provides several entry points to create {@link ArchModule modules} from a set of {@link JavaClass classes} in a convenient way:<br>
+ * <ul>
+ *     <li>{@link #defineBy(IdentifierAssociation)} - the most generic API allowing a lot of flexibility</li>
+ *     <li>{@link #defineByPackages(String)} - an API similar to {@link Slices#matching(String)}</li>
+ * </ul>
```

Comment:
Should `{@link #defineByRootClasses(Predicate)}`, `{@link #defineByAnnotation(Class)}`, `{@link #defineByAnnotation(Class, Function)}` also be mentioned here?

Reply:
Yes, I wanted to extend the docs in every further commit :see_no_evil: